"SPA" компонент на Vue

kostyanvini opened the pull request.

kostyanvini committed:
"SPA" компонент на Vue

nook-ru commented on local/components/kostyanvi/news/class.php:
```diff
@@ -0,0 +1,66 @@
+	public function getNewsByIDAction()
+	{
+		if (\Bitrix\Main\Loader::includeModule('iblock')) {
+			$newsDetailInfo = [];
```
Это лишняя переменная, чуть ниже она переопределяется

nook-ru commented on local/components/kostyanvi/news/class.php:
```diff
@@ -0,0 +1,66 @@
+	{
+		if (\Bitrix\Main\Loader::includeModule('iblock')) {
+			$newsDetailInfo = [];
+			$request = \Bitrix\Main\Context::getCurrent()->getRequest();
```
Запрос из компонента можно получить как `$this->getRequest()`, а ещё лучше использовать т.н. autowiring. Параметры запроса подставляются в параметры метода контроллера по названию.

Можно заменить сигнатуру метода на

```php
public function getNewsByIDAction(int $id)
```

Тогда в аргументе id окажется одноименного параметра запроса.

nook-ru commented on local/components/kostyanvi/news/class.php:
```diff
@@ -0,0 +1,66 @@
+
+			$detailEl = \Bitrix\Iblock\ElementTable::getById($request->get('id'));
+
+			$newsDetailInfo = $detailEl->fetch();
```
Тоже лишняя переменная:

```suggestion
			$newsDetailInfo = \Bitrix\Iblock\ElementTable::getById($request->get('id'))->fetch();
```

nook-ru commented on local/components/kostyanvi/news/class.php:
```diff
@@ -0,0 +1,66 @@
+
+	public function getNewsByIDAction()
+	{
+		if (\Bitrix\Main\Loader::includeModule('iblock')) {
```
Чтобы код был читабельнее, лучше не делать больших ветвлений. Лучше проверить на наличие ошибок и выбросить исключение или вернуть ошибку.

Тогда else окажется не нужен.

И повторять один и тот же код в разных методах не стоит, можно подключить необходимые модули в конструкторе. В ответ на запросы к экшенам компонент вернет json с ошибкой.

nook-ru commented on local/components/kostyanvi/news/class.php:
```diff
@@ -0,0 +1,66 @@
+
+			$newsDetailInfo['PREVIEW_PICTURE'] = CFile::GetFileArray($newsDetailInfo['PREVIEW_PICTURE']);
+
+			return Json::encode($newsDetailInfo);
```
Это лишнее, можешь вернуть массив, битрикс в ответе представить его как json.

nook-ru commented on local/components/kostyanvi/news/class.php:
```diff
@@ -0,0 +1,66 @@
+			$dbNewsResult = \Bitrix\Iblock\ElementTable::getList([
+				'select' => ['*'],
+				'filter' => [
+					'IBLOCK_ID' => 1
```
Такое обычно выносят в параметры компонента, ну или хотя бы в константы, если не планируется делать это настраиваемым через админку.

nook-ru replied:
Наверное, если вынести подключение модуля в конструктор, метод можно сократить:

```php
return array_map(
	function ($news) {
		$news['PREVIEW_PICTURE'] = CFile::GetFileArray($news['PREVIEW_PICTURE']);
		return $news;
	},
	\Bitrix\Iblock\ElementTable::getList([
		'select' => ['*'],
		'filter' => [
			'IBLOCK_ID' => 1,
		]
	])->fetchAll()
);
```

nook-ru replied:
Или так:

```php
return array_map(
	function ($news) {
		$news['PREVIEW_PICTURE'] = CFile::GetFileArray($news['PREVIEW_PICTURE']);
		return $news;
	},
	\Bitrix\Iblock\ElementTable::query()
		->setSelect(['*'])
		->where('IBLOCK_ID', 1)
		->fetchAll()
);
```

nook-ru commented on local/components/kostyanvi/news/templates/news/component_epilog.php:
```diff
@@ -0,0 +1,5 @@
+
+\Bitrix\Main\UI\Extension::load('ui.vue');
+\Bitrix\Main\UI\Extension::load('ui.vue.router');
+\Bitrix\Main\UI\Extension::load('ui.vue.vuex');
```
Можно передавать массив из нескольких расширений:

```suggestion
\Bitrix\Main\UI\Extension::load([
  'ui.vue',
  'ui.vue.router',
  'ui.vue.vuex'
]);
```

nook-ru commented on local/components/kostyanvi/news/templates/news/template.php:
```diff
@@ -0,0 +1,16 @@
+<? if (!defined("B_PROLOG_INCLUDED") || B_PROLOG_INCLUDED !== true) {
+	die();
+}
+const VUEJS_DEBUG = true;
```
Это не в шаблоне, а где-нибудь в init.php или dbconn.php должно задаваться

nook-ru commented on local/components/kostyanvi/news/templates/news/vue/src/components/News.vue:
```diff
@@ -0,0 +1,81 @@
+        if (this.$store.state.news.length) {
+            this.news = this.$store.state.news;
+        } else {
+            BX.ajax.runComponentAction('kostyanvi:news', 'getAllNews', {
```
Название компонента мы обычно в шаблоне компонента БУС получаем, и передаем затем во входные параметры компонента vue.

```php
<?php
$vueParams = \Bitrix\Main\Web\Json::encode([
  'componentName' => $this->getComponent()->getName(),
], JSON_UNESCAPED_UNICODE);
?>

<div id="news"></div>

<script>
	BX.ready(function () {
		BX.NewsManager.init('#news', <?=$vueParams?>);
	});
</script>
```

nook-ru replied:
Когда в запросе с фронта к действию контроллера нужно безопасно пробросить входные параметры компонента БУС, вместе  `componentName` потребуется передать [signedParameters](https://dev.1c-bitrix.ru/learning/course/index.php?COURSE_ID=43&LESSON_ID=2192).

nook-ru commented on local/components/kostyanvi/news/templates/news/vue/src/components/NewsDetail.vue:
```diff
@@ -0,0 +1,73 @@
+                }
+            ).then((responce => {
+                try {
+                    const result = JSON.parse(responce.data);
```
Здесь JSON декодировать тоже не придется, если убрать Json::encode в ответе контроллера. Битрикс сам об этом позаботится.

nook-ru commented on local/components/kostyanvi/news/templates/news/vue/src/components/NewsDetail.vue:
```diff
@@ -0,0 +1,73 @@
+                    this.newsDetailInfo = result;
+                    this.$store.commit('registerNewNewsCache', result);
+                } catch (e) {
+                    console.warn('Не удалось получить данные');
```
Тут скорее не удается json декодировать, исключения и ошибки бека можно и нужно обрабатывать на фронте через .catch() прописа, что возвращает `runComponentAction`

kostyanvini committed:
Рефакторинг компонента
• Добавлена "поддержка" ЧПУ
• Добавлен компонент прелоадера
• Исправлены методы контроллера
• ID инфоблока вынесен в параметр вызова

alekseevamaria commented on local/components/kostyanvi/news/templates/news/template.php:
```diff
@@ -0,0 +1,19 @@
+<? if (!defined("B_PROLOG_INCLUDED") || B_PROLOG_INCLUDED !== true) {
+	die();
+}
+$this->setFrameMode(false);
```
Почему false, а не true?

alekseevamaria commented on local/components/kostyanvi/news/templates/news/vue/src/components/News.vue:
```diff
@@ -0,0 +1,83 @@
+            this.news = this.$store.state.news;
+            this.isLoaded = false
+        } else {
+            BX.ajax.runComponentAction(this.$root.componentName, 'getAllNews', {
```
Вместо доступа к данным родительского компонента через this.$root не верно  делать.
Для этого есть [props](https://v2.vuejs.org/v2/guide/components-props.html)

alekseevamaria replied:
Название метода  'getAllNews' тоже лучше получать из props.
При изменении в бэке, придется фронт пересобирать. А так из настроек попадет.

alekseevamaria commented on local/components/kostyanvi/news/templates/news/vue/src/components/News.vue:
```diff
@@ -0,0 +1,83 @@
+                    const news = responce.data;
+                    this.news = news;
+                    this.$store.commit('setAllNews', news);
+                    setTimeout(() => this.isLoaded = false, 1000)
```
Зачем задержка для прелоудера?

nook-ru commented on local/components/kostyanvi/news/.description.php:
```diff
@@ -6,7 +6,7 @@
 	"NAME" => GetMessage('COMPONENT_DESCRIPTION_NEWS_NAME'),
 	"DESCRIPTION" => GetMessage('COMPONENT_DESCRIPTION_NEWS_DESCRIPTION'),
-	"SORT" => 10,
+	"COMPLEX" => "Y",
```
Понятие комплексных компонентов довольно условное, и этот ключ влияет по-моему только на положение в дереве компонентов в виз. редакторе.

К тому же у тебя роутинг на фронте, так что БУСу пофиг на это.

kostyanvini replied:
Понял, поправлю

nook-ru commented on local/components/kostyanvi/news/class.php:
```diff
@@ -12,55 +12,63 @@ public function configureActions()
+		try {
+			\Bitrix\Main\Loader::includeModule('iblock');
+		} catch (Exception $err) {
+			return $err;
```
Если честно, даже не задумывался что будет, если вернуть что-то из конструктора. По-моему ничего)

Исключения там вроде и не будет, у тебя же includeModule, а не requireModule. Думаю можно оставить исключение без обработки, сразу видно будет, что что-то конкретно сломалось.

nook-ru commented on local/components/kostyanvi/news/class.php:
```diff
@@ -12,55 +12,63 @@ public function configureActions()
-			$newsDetailInfo['PREVIEW_PICTURE'] = CFile::GetFileArray($newsDetailInfo['PREVIEW_PICTURE']);
+	public function onPrepareComponentParams($arParams)
+	{
+		return parent::onPrepareComponentParams($arParams);
```
Здесь обычно валидируются и преобразуются параметры компонента, просто вызывать метод родительского класса нет смысла - там нет ничего.

kostyanvini replied:
Оставил на будущее. Но да, так как сейчас параметры я не валидирую, то и метод не нужен

kostyanvini committed:
Рефакторинг компонента #2
• Вынесены методы получения данных в пропсы компонента
• Убраны таймауты прелоадера

kostyanvini committed:
Пагинация

alekseevamaria commented on local/components/kostyanvi/news/templates/news/vue/src/components/News.vue:
```diff
@@ -50,9 +53,26 @@ export default {
+                    this.news = responce.data;
+                    this.isLoaded = false;
+                })
+            ).catch(err => console.warn(err));
```
Тут тоже надо  this.isLoaded = false; делать.
Иначе прелоудер так и останется

alekseevamaria commented on local/components/kostyanvi/news/class.php:
```diff
@@ -71,4 +73,30 @@ function ($news) {
+
+		$news = \Bitrix\Iblock\ElementTable::getList(
+			array(
+				"filter" => ["=IBLOCK_ID" => 1],
```
Вместо 1 - надо использовать занчение из параметров компонента $this->arParams['IBLOCK_ID']

alekseevamaria commented on local/components/kostyanvi/news/class.php:
```diff
@@ -71,4 +73,30 @@ function ($news) {
 		);
 	}
+
+	public function getPageNewsAction(int $page = 1)
```
При передаче номера страницы, которой не существует, компонент ведет себя корректно? Не выдает данные первой страницы?

alekseevamaria commented on local/components/kostyanvi/news/templates/news/vue/src/components/Pagination.vue:
```diff
@@ -0,0 +1,43 @@
+        <a href="javascript:void(0)" class="pagination__item" @click="$emit('pageChanged', 2)">2</a>
+        <a href="javascript:void(0)" class="pagination__item" @click="$emit('pageChanged', 3)">3</a>
+        <a href="javascript:void(0)" class="pagination__item" @click="$emit('pageChanged', 4)">4</a>
+        <a href="javascript:void(0)" class="pagination__item" @click="$emit('pageChanged', 5)">5</a>
```
1. Как пользователю понять, какая текущая страница?
2. Список страниц  (их количество надо получать из данных компонента из class.php).
Если данных в инфоблоке будет не на 5 страниц, на одну, или 10 например
3. Если нужно именно 5 ссылок, то лучше циклом вывести данный блок через [v-for](https://ru.vuejs.org/v2/guide/list.html)
4. Можно еще реализовать, чтобы можно было открыть ссылку на нужную страницу в новой вкладке

alekseevamaria commented on local/components/kostyanvi/news/templates/news/vue/src/components/News.vue:
```diff
@@ -12,14 +12,17 @@
                 <div class="news__item-text" v-html="el.PREVIEW_TEXT"></div>
             </div>
         </div>
+        <Pagination @pageChanged="changePage"></Pagination>
```
Тут надо передать в компонент пагинации данные по пагинации из $arResult компонента, чтобы ссылок было столько, сколько нужно

kostyanvini committed:
Правки пагинации
Добавлена страница 404
Добавлена страница со списком всех новостей
